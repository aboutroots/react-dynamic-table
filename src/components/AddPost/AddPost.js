// @flow
import React from 'react';
import moment from 'moment';

import { connect } from 'react-redux';
import { addPost } from '../../actions/posts';

import type { postType } from '../../flow/exports';

type Props = {
  addPost: (post: postType) => void,
  afterAdd: () => void,
};

type State = {
  id: string,
  username: string,
  postTitle: string,
  views: string,
  likes: string,
  createdAt: string,
};
export class AddPost extends React.Component<Props, State> {
  state = {
    id: '',
    username: '',
    postTitle: '',
    views: '',
    likes: '',
    createdAt: '',
  };

  handleSubmit = (e: SyntheticEvent<HTMLFormElement>) => {
    e.preventDefault();
    let { id, username, postTitle, likes, views, createdAt } = this.state;

    // parse string values into correct types
    createdAt = moment(createdAt).valueOf();
    id = parseInt(id);
    likes = parseInt(likes);
    views = parseInt(views);

    const post = { id, username, postTitle, likes, views, createdAt };
    if (this.validatePost(post)) {
      this.props.addPost(post);
      this.props.afterAdd();
    }
  };

  handleChange = (inputName: string, inputValue: string) => {
    this.setState(state => ({
      ...state,
      [inputName]: inputValue,
    }));
  };

  validatePost = ({
    id,
    username,
    postTitle,
    likes,
    views,
    createdAt,
  }: postType) => {
    let numberTypesFailed = [id, likes, views].some(
      val => Number.isNaN(val) || val < 0
    );
    let stringTypesFailed = [username, postTitle].some(
      val => val.trim() === ''
    );
    let dateNumberFailed = Number.isNaN(createdAt);

    return numberTypesFailed || stringTypesFailed || dateNumberFailed
      ? false
      : true;
  };

  render() {
    return (
      <div className="AddPost">
        <form onSubmit={this.handleSubmit} className="AddPost__form">
          <input
            placeholder="#ID"
            type="number"
            value={this.state.id}
            onChange={e => this.handleChange('id', e.target.value)}
            className="AddPost__form-element inpt"
          />
          <input
            placeholder="User Name"
            type="text"
            value={this.state.username}
            onChange={e => this.handleChange('username', e.target.value)}
            className="AddPost__form-element inpt"
          />
          <input
            placeholder="Post Title"
            type="text"
            value={this.state.postTitle}
            onChange={e => this.handleChange('postTitle', e.target.value)}
            className="AddPost__form-element inpt"
          />
          <input
            placeholder="Views"
            type="number"
            value={this.state.views}
            onChange={e => this.handleChange('views', e.target.value)}
            className="AddPost__form-element inpt"
          />
          <input
            placeholder="Likes"
            type="number"
            value={this.state.likes}
            onChange={e => this.handleChange('likes', e.target.value)}
            className="AddPost__form-element inpt"
          />
          <input
            placeholder="Created At"
            type="date"
            value={this.state.createdAt}
            onChange={e => this.handleChange('createdAt', e.target.value)}
            className="AddPost__form-element inpt"
          />
          <button
            type="submit"
            className="AddPost__form-element AddPost__button btn"
          >
            Add Post
          </button>
        </form>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addPost: post => dispatch(addPost(post)),
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddPost);
