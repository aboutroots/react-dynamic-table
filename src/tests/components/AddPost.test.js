import React from 'react';
import { shallow } from 'enzyme';
import { AddPost } from '../../components/AddPost/AddPost';
import moment from 'moment';

let wrapper, addPost, afterAdd;

beforeEach(() => {
  addPost = jest.fn();
  afterAdd = jest.fn();
  wrapper = shallow(<AddPost addPost={addPost} afterAdd={afterAdd} />);
});

test('should render AddPost properly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should call addPost on submit', () => {
  wrapper.setState({
    id: '1',
    username: 'test',
    postTitle: 'test',
    views: '1',
    likes: '1',
    createdAt: moment('1970-01-01', 'YYYY-MM-DD').valueOf(),
  });
  wrapper.find('form').simulate('submit', { preventDefault: () => {} });
  expect(addPost).toHaveBeenLastCalledWith({
    id: 1,
    username: 'test',
    postTitle: 'test',
    views: 1,
    likes: 1,
    createdAt: moment('1970-01-01', 'YYYY-MM-DD').valueOf(),
  });
});

test('should call afterAdd on submit', () => {
  wrapper.setState({
    id: '1',
    username: 'test',
    postTitle: 'test',
    views: '1',
    likes: '1',
    createdAt: moment('1970-01-01', 'YYYY-MM-DD').valueOf(),
  });
  wrapper.find('form').simulate('submit', { preventDefault: () => {} });
  expect(afterAdd).toHaveBeenCalled();
});

test('should not add post with any empty value', () => {
  let defaultState = {
    id: '1',
    username: 'test',
    postTitle: 'test',
    views: '1',
    likes: '1',
    createdAt: moment('1970-01-01', 'YYYY-MM-DD').valueOf(),
  };
  Object.keys(defaultState).forEach(key => {
    wrapper.setState({
      ...defaultState,
      [key]: '',
    });
    wrapper.find('form').simulate('submit', { preventDefault: () => {} });
    expect(addPost).not.toHaveBeenCalled();
  });
});
