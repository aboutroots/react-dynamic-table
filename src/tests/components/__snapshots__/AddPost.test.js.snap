// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddPost properly 1`] = `
<div
  className="AddPost"
>
  <form
    className="AddPost__form"
    onSubmit={[Function]}
  >
    <input
      className="AddPost__form-element"
      onChange={[Function]}
      placeholder="#ID"
      type="number"
      value=""
    />
    <input
      className="AddPost__form-element"
      onChange={[Function]}
      placeholder="User Name"
      type="text"
      value=""
    />
    <input
      className="AddPost__form-element"
      onChange={[Function]}
      placeholder="Post Title"
      type="text"
      value=""
    />
    <input
      className="AddPost__form-element"
      onChange={[Function]}
      placeholder="Views"
      type="number"
      value=""
    />
    <input
      className="AddPost__form-element"
      onChange={[Function]}
      placeholder="Likes"
      type="number"
      value=""
    />
    <input
      className="AddPost__form-element"
      onChange={[Function]}
      placeholder="Created At"
      type="date"
      value=""
    />
    <button
      className="AddPost__form-element AddPost__button"
      type="submit"
    >
      Add Post
    </button>
  </form>
</div>
`;
